### PR TITLE
Allow user to set leaf position for `PathVar`

### PR DESCRIPTION
**Pull request by tsunrise**
## Description

Currently, for `PathVar`, the location information is encoded by `path: Vec<Boolean>` and `leaf_is_right_child: Boolean` (chaining those two get the big endian encoding of index of leaf). 

There are indeed many circumstances where we also have constraints on leaf position (take IOP verifier's query as an example). For now, `path` and `leaf_is_right_child` are just variables created from `Boolean::new_variable`, and user do not have access to such those variables. In this commit, user will be able to: 
- get the leaf index variable in `PathVar` and constrain them
- replace the leaf index variable in `PathVar` with their own variable (possibly already with some constraints)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by daira** [comment]
Leaf position is also used in the Sapling circuit.

**Comment by tsunrise** [comment]
@weikengchen Does code look good? If so we can merge it

File: src/merkle_tree/constraints.rs [hunk 1]
```diff
@@ -149,6 +149,34 @@ where
...
+    /// to check the correctness leaf position.
+    /// * `leaf_index`: leaf index encoded in little-endian format
+    #[tracing::instrument(target = "r1cs", skip(self))]
+    pub fn set_leaf_position(&mut self, leaf_index: Vec<Boolean<ConstraintF>>) {
+        // convert leaf_index to path:
+        // first remove the element (which represent a leaf), pad it to appropriate length
+        // and reverse it to become big endian.
+        // Also, set `leaf_is_right_child` accordingly.
+        let mut path = leaf_index;
+        let leaf_is_right_child = path.remove(0);
```
Comment (Pratyush) [inline src/merkle_tree/constraints.rs:162]:
```suggestion
		// If leaves are numbered left-to-right starting from zero,
		// then all left children have odd positions, while all
		// right children have even positions.
        let leaf_is_right_child = path.remove(0);
```

File: src/merkle_tree/constraints.rs [hunk 2]
```diff
@@ -149,6 +149,34 @@ where
...
+    /// Set the leaf index of the path to a given value. Verifier can use function before calling `verify`
+    /// to check the correctness leaf position.
+    /// * `leaf_index`: leaf index encoded in little-endian format
+    #[tracing::instrument(target = "r1cs", skip(self))]
+    pub fn set_leaf_position(&mut self, leaf_index: Vec<Boolean<ConstraintF>>) {
+        // convert leaf_index to path:
+        // first remove the element (which represent a leaf), pad it to appropriate length
+        // and reverse it to become big endian.
+        // Also, set `leaf_is_right_child` accordingly.
+        let mut path = leaf_index;
```
Comment (Pratyush) [inline src/merkle_tree/constraints.rs:161]:
```suggestion
		
		// The path to a leaf is described by the branching
		// decisions taken at each node. This corresponds to the position
		// of the leaf.
        let mut path = leaf_index;
```

**Review by Pratyush — APPROVED** [review]

**Comment by Pratyush** [comment]
The changes look good, modulo the two comments. Sorry for the delay in reviewing!

File: src/merkle_tree/constraints.rs (hunk 2 above)
Comment (tsunrise) [reply src/merkle_tree/constraints.rs:160]: fixed

File: src/merkle_tree/constraints.rs (hunk 1 above)
Comment (tsunrise) [reply src/merkle_tree/constraints.rs:165]: fixed

**Review by tsunrise — COMMENTED** [review]
Thanks for the review! Fixed

**Pull request merged**